Ember Release blog post v4.5 (#1162)
ember-learn/ember-blog: content/ember-4-5-released.md

- [The Ember Project](https://blog.emberjs.com/ember-project-at-2-0/)
- [Ember LTS Releases](https://blog.emberjs.com/announcing-embers-first-lts/)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 4.5

Ember.js 4.5 is an incremental, backwards compatible release of Ember with bug fixes, performance improvements, and minor deprecations.

#### Bug Fixes

Ember.js 4.5 introduced 0 bug fixes.

#### Features

Ember.js 4.5 introduced 2 new features.

1. Plain function as helpers
2. A new `renderSettled` test helper

##### 1. Plain functions as helpers

You can now use plain functions as helpers in your component templates. This helps make the relationship between Ember component templates and their JavaScript class more intuitive.

For example, here we create a method `double` and use it directly in a template:

```js
// my-component.js

import Component from '@glimmer/component';

export default class MyComponent extends Component {
  double = num => num * 2;
}
```

```hbs
// my-component.hbs

{{this.double 2}}

<SomeComponent @foo={{this.double 2}} />
```

> Previously, you would need to [write a separate helper](https://guides.emberjs.com/release/components/helper-functions/#toc_writing-a-helper-function) in order to accomplish this.

Reviewer: This isn't totally true. You could do the following, as of 3.25:

```js
export default class MyComponent extends Component {
  double = helper(([num]) => num * 2);
}
```

Reviewer: Some common feedback I've gotten from co-workers, is that they're blown away by the DX of "anything as values from anywhere" and how it's a massive quality of life improvement.

They ask for guides / docs on this behavior, and I'm just like :shrug: idk what to tell them, because idk... _you *just* assign the value_.

Maybe since the guides are somewhat focused on globally available everything, they think local stuff isn't possible? idk.

Maybe there is a way to address? idk if this blog post is the right place to mention all this, but. yeah. lots of idk :upside_down_face:

Author: Thank you, I’ll incorporate this!

Author: I dropped “separate” and kept the link to the guides as-is. Does that seem ok?

Reviewer: I think my gut recation is to the word "need"?

maybe 

>  Even though local helpers have been supported since 3.25, the documentation says to [write a separate helper](https://guides.emberjs.com/release/components/helper-functions/#toc_writing-a-helper-function) in order to accomplish this. 

maybe?

Author: I wonder if we inadvertently over-corrected for people being confused by data passing. I think we can address this confusion in the component guides, potentially.

Reviewer: yeah potentially -- related, I ran in to someone who thought accessing args during initialization was impossible and sent them [this](https://limber.glimdown.com/?t=%23%20Limber%20Editor%0A%0AAccessing%20%60%40arg%60s%20during%20initialization.%0A%0A%60%60%60gjs%20live%20preview%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aclass%20Child%20extends%20Component%20%7B%0A%20%20%2F%2F%20will%20not%20receive%20arg%20updates%20%0A%20%20%2F%2F%20(good%20for%20letting%20the%20user%20override%20passed%20in%20state)%0A%20%20%40tracked%20dedupedValue%3B%0A%0A%20%20constructor(owner%2C%20args)%20%7B%0A%20%20%20%20super(owner%2C%20args)%3B%0A%0A%20%20%20%20%2F%2F%20setup%20here%0A%20%20%20%20this.dedupedValue%20%3D%20args.value%3B%0A%20%20%7D%0A%0A%20%20update%20%3D%20(event)%20%3D%3E%20this.dedupedValue%20%3D%20event.target.value%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%7B%7Bthis.dedupedValue%7D%7D%0A%0A%20%20%20%20%3Cbr%3E%0A%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20Local%20change%0A%20%20%20%20%20%20%3Cinput%20%7B%7Bon%20%27input%27%20this.update%7D%7D%20class%3D%22border%22%3E%0A%20%20%20%20%3C%2Flabel%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%0A%0A%3Ctemplate%3E%0A%20%20%3CChild%20%40value%3D%22hello%20there%22%20%2F%3E%0A%3C%2Ftemplate%3E%0A%60%60%60%0A)

It seems some folks somehow think data passing special and works different from what they expect.

Thankfully, since 3.25, all data/value passing works the same as it does in JS.

Author: I applied your wording. Thanks!

We're working on updating the Guides to cover this pattern.
For background, check out [RFC 756](https://rfcs.emberjs.com/id/0756-helper-default-manager), which designed this feature.

> Also, keep your eyes on this blog: we will have a dedicated blog post with a deep dive on this new capability in the next week or two!

Reviewer: :partying_face:

This release kicks off the 4.5 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

You can read more about our general release process here:

- [Release Dashboard](http://emberjs.com/releases/)
- [The Ember Release Cycle](https://blog.emberjs.com/new-ember-release-process/)
- [The Ember Project](https://blog.emberjs.com/ember-project-at-2-0/)
- [Ember LTS Releases](https://blog.emberjs.com/announcing-embers-first-lts/)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 4.5

Ember.js 4.5 is an incremental, backwards compatible release of Ember with bug fixes, performance improvements, and minor deprecations.

#### Bug Fixes

Ember.js 4.5 introduced 0 bug fixes.

#### Features

Ember.js 4.5 introduced 2 new features.

1. Plain function as helpers

> 2. `renderSettled` is now public API

Reviewer 2: Tweaking the language here just a bit for "interesting-ness":

```suggestion
2. A new `renderSettled` test helper
```

##### 1. Plain functions as helpers

You can now use plain functions as helpers in your component templates. This helps make the relationship between Ember component templates and their JavaScript class more intuitive.

> The full documentation can be found in [TODO link to guides]()

Reviewer 2: Maybe let's remove this sentence from here, and then add some text below after the example?

```suggestion
```

For example, here we create a method `double` and use it directly in a template:

```js
// my-component.js

import Component from '@glimmer/component';

export default class MyComponent extends Component {
  double = num => num * 2;
}
```

```hbs
// my-component.hbs

{{this.double 2}}

<SomeComponent @foo={{this.double 2}} />
```

Previously, you would need to [write a separate helper](https://guides.emberjs.com/release/components/helper-functions/#toc_writing-a-helper-function) in order to accomplish this.

> This feature was made possible when [RFC 756](https://rfcs.emberjs.com/id/0756-helper-default-manager) was implemented.

Reviewer 2: ```suggestion
We're working on updating the Guides to cover this pattern.
For background, check out [RFC 756](https://rfcs.emberjs.com/id/0756-helper-default-manager), which designed this feature.
Also, keep your eyes on this blog: we will have a dedicated blog post with a deep dive on this new capability in the next week or two!
```

> ##### 2. `renderSettled` is now public API

Reviewer 2: ```suggestion
##### 2. A new `renderSettled` test helper
```

> `renderSettled` is a method you can use in your application tests. For example, in a test, you might want to update some tracked state and then run some assertions after rendering has completed. The `renderSettled` method returns a promise which fulfills when rendering has completed.
> 
> ```
> import { renderSettled } from '@ember/renderer';
> ```
> 
> `renderSettled` is one chapter in some upcoming improvements to Ember's testing story. The ongoing work and goals are described in [RFC 785](https://rfcs.emberjs.com/id/0785-remove-set-get-in-tests).

Reviewer 2: ````suggestion
Under the hood, Ember's tests use a "test waiters" system to allow you to control the flow of your tests in terms of the actual framework behavior.
That way that your tests match exactly what the app does at runtime.
However, making this work depends on Ember providing all the necessary hooks for test helpers to use, and there was one significant missing public API.
You could wait for *all* of the test waiters to finish with `settled`, but there was no public way to wait for *just rendering* to finish.
For example, you might want to wait for rendering to finish but *not* for an Ember Data `save` operation to finish, as part of testing a loading screen.

Ember 4.5 introduces a new function, `renderSettled`, as a public way for test helpers to interact with the rendering phase of the application.
`renderSettled` returns returns a promise which fulfills as soon as rendering has completed.
It can be used in any rendering or application test.
(It also works in other tests where you set up the rendering hooks manually, but this is unusual!)

```
import { renderSettled } from '@ember/renderer';
```

An upcoming release of `@ember/test-helpers` will take advantage of this to provide a new `await rerender()` helper.
For more details, and how this fits into improvements to Ember's testing story, see [RFC 785](https://rfcs.emberjs.com/id/0785-remove-set-get-in-tests).
````

---
title: Ember 4.5 Released
authors:
  - jen-weber
  - chris-krycho
date: 2022-07-12T00:00:00.000Z
tags:
  - releases
  - '2022'
  - version-4-x
---

Today the Ember project is releasing version 4.5 of Ember.js, Ember Data, and Ember CLI.

> 

Reviewer 2: ```suggestion

Version 4.4 of Ember is now promoted to LTS (Long Term Support).
An LTS version of Ember continues to receive security updates for 9 release cycles (54 weeks) and bugfixes for 6 cycles (36 weeks).
LTS releases typically occur every four minor versions.
The previous LTS version of Ember was 3.28.

```

This release kicks off the 4.5 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

> 

Reviewer 3: The release of 4.5 is also when 4.4 is promoted to LTS.

- [The Ember Project](https://blog.emberjs.com/ember-project-at-2-0/)
- [Ember LTS Releases](https://blog.emberjs.com/announcing-embers-first-lts/)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 4.5

Ember.js 4.5 is an incremental, backwards compatible release of Ember with bug fixes, performance improvements, and minor deprecations.

#### Bug Fixes

Ember.js 4.5 introduced 0 bug fixes.

#### Features

Ember.js 4.5 introduced 2 new features.

1. Plain function as helpers
2. A new `renderSettled` test helper

##### 1. Plain functions as helpers

You can now use plain functions as helpers in your component templates. This helps make the relationship between Ember component templates and their JavaScript class more intuitive.

For example, here we create a method `double` and use it directly in a template:

```js
// my-component.js

import Component from '@glimmer/component';

export default class MyComponent extends Component {
  double = num => num * 2;
}
```

```hbs
// my-component.hbs

{{this.double 2}}

<SomeComponent @foo={{this.double 2}} />
```

> Previously, documentation said to [write a separate helper](https://guides.emberjs.com/release/components/helper-functions/#toc_writing-a-helper-function) in order to accomplish this.

Reviewer 2: ```suggestion
Previously, you could define this locally, but still had to [use the `helper()`](https://guides.emberjs.com/release/components/helper-functions/#toc_writing-a-helper-function) function to accomplish this.
```

##### 2. A new `renderSettled` test helper

Under the hood, Ember's tests use a "test waiters" system to allow you to control the flow of your tests in terms of the actual framework behavior.
That way that your tests match exactly what the app does at runtime.
However, making this work depends on Ember providing all the necessary hooks for test helpers to use, and there was one significant missing public API.
You could wait for *all* of the test waiters to finish with `settled`, but there was no public way to wait for *just rendering* to finish.
For example, you might want to wait for rendering to finish but *not* for an Ember Data `save` operation to finish, as part of testing a loading screen.

Ember 4.5 introduces a new function, `renderSettled`, as a public way for test helpers to interact with the rendering phase of the application.
`renderSettled` returns returns a promise which fulfills as soon as rendering has completed.
It can be used in any rendering or application test.
(It also works in other tests where you set up the rendering hooks manually, but this is unusual!)

```js
import { renderSettled } from '@ember/renderer';
```

> An upcoming release of `@ember/test-helpers` will take advantage of this to provide a new `await rerender()` helper.

Reviewer 3: This was released in v2.8.0 to implement the RFC 785 https://github.com/emberjs/ember-test-helpers/blob/master/CHANGELOG.md#v280-2022-05-17

Reviewer 2: Oh, awesome; I didn't realize it was out already!

Reviewer 2: ```suggestion
An recent release of `@ember/test-helpers`, v2.8.0, takes advantage of this to provide a new `await rerender()` helper.
```

> An recent release of `@ember/test-helpers`, v2.8.0, takes advantage of this to provide a new `await rerender()` helper.

Reviewer 2: ```suggestion
An recent release of `@ember/test-helpers`, [v2.8.0](https://github.com/emberjs/ember-test-helpers/blob/master/CHANGELOG.md#v280-2022-05-17), takes advantage of this to provide a new `await rerender()` helper.
```

Due to low availability for the Ember Data team this cycle, Ember Data does not have a new release, so it remains at 4.4.

Please see the [Ember 4.4 release blog post](https://blog.emberjs.com/ember-released-4-4) for details about `v4.4` of Ember Data.

---

## Ember CLI

Ember CLI is the command line interface for managing and packaging Ember.js applications.

### Upgrading Ember CLI

You may upgrade Ember CLI using the `ember-cli-update` project:

```bash
npx ember-cli-update
```

This utility will help you to update your app or addon to the latest Ember CLI version. You will probably encounter merge conflicts, in which the default behavior is to let you resolve conflicts on your own. For more information on the `ember-cli-update` project, see [the GitHub README](https://github.com/ember-cli/ember-cli-update).

While it is recommended to keep Ember CLI versions in sync with Ember and Ember Data, this is not required. After updating ember-cli, you can keep your current version(s) of Ember or Ember Data by editing `package.json` to revert the changes to the lines containing `ember-source` and `ember-data`.

### Changes in Ember CLI 4.5

> This Ember CLI release drops support for Node 12. Node 12 reached "end of life"status (it no longer receives security updates) in April 2022.

Reviewer 3: ```suggestion
This Ember CLI release drops support for Node 12. Node 12 reached "end of life" status (it no longer receives security updates) in April 2022.
```
